File: catkin_ws/src/asclinic_pkg/src/drivers/src/teensy41/teensy41.cpp
#include "teensy41/teensy41.h"

Teensy41::Teensy41()
{
	strcpy(this->device_name, "/dev/ttyACM0");
	this->open_device();
}

Teensy41::Teensy41(const char *device_name)
{
	if ((int)strlen(device_name) < TEENSY_MAX_DEVICE_NAME_LENGTH)
	{
		strcpy(this->device_name, device_name);
	}
	else
	{
		// Inform the user
		printf("FAILED to initialise driver becuse provided device_name is too long.\n");
		printf("> device_name = %s\n", device_name);
		printf("> strlen(device_name) = %d, MAX_DEVICE_NAME_LENGTH = %d\n",
			(int)strlen(device_name),
			TEENSY_MAX_DEVICE_NAME_LENGTH);
		printf("> Defaulting instead to /dev/ttyACM0\n");
		// Set the device name to a default
		strcpy(this->device_name, "/dev/ttyACM0");
	}
	this->open_device();
}

void Teensy41::open_device()
{
	this->file_descriptor = open(this->device_name, O_RDWR);
	if (this->file_descriptor == -1)
	{
		perror(this->device_name);
	}
}

bool Teensy41::close_device()
{
	if (close(this->file_descriptor) == -1)
	{
		perror(this->device_name);
		return false;
	}
	return true;
}

bool Teensy41::get_motors_displacement(int *motor1, int *motor2)
{
	write(this->file_descriptor, &TEENSY_REQUEST_DISPLACEMENT_BYTE, 1);
	size_t returned_packet_size = read(
		this->file_descriptor,
		this->returned_buffer,
		TEENSY_RETURNED_BUFFER_SIZE
	);
	this->returned_buffer[returned_packet_size] = '\0';
	if (sscanf(this->returned_buffer, "<%d,%d>", motor1, motor2) != 2)
	{
		return false;
	}
	return true;
}

bool Teensy41::get_motors_displacement_with_timestamp(
	int *disp1, int *disp2, unsigned int *timestamp)
{
	write(this->file_descriptor, &TEENSY_REQUEST_DISPLACEMENT_WITH_TIMESTAMP_BYTE, 1);
	size_t returned_packet_size = read(
		this->file_descriptor,
		this->returned_buffer,
		TEENSY_RETURNED_BUFFER_SIZE
	);
	this->returned_buffer[returned_packet_size] = '\0';
	if (sscanf(this->returned_buffer, "{%d,%d,%u}", disp1, disp2, timestamp) != 3)
	{
		return false;
	}
	return true;
}

bool Teensy41::get_motors_speed_rpm(float *speed1, float *speed2, int delta_t_in_us)
{
	int disp1, disp2;
	// The first ``get_motors_displacement`` set the wheel position to 0
	if (!this->get_motors_displacement(&disp1, &disp2))
	{
		return false;
	}
	usleep(delta_t_in_us);
	if (!this->get_motors_displacement(&disp1, &disp2))
	{
		return false;
	}
	// speed = disp1 / 3200.0 / (delta_t_in_us / 60000000);
	*speed1 = ((float)disp1) * 0.0003125 / (delta_t_in_us * 1.667e-8);
	*speed2 = ((float)disp2) * 0.0003125 / (delta_t_in_us * 1.667e-8);
	return true;
}

bool Teensy41::get_motors_speed_rpm_directly_from_teensy(float *speed1, float *speed2)
{
	write(this->file_descriptor, &TEENSY_REQUEST_SPEED_BYTE, 1);
	size_t returned_packet_size = read(
		this->file_descriptor,
		this->returned_buffer,
		TEENSY_RETURNED_BUFFER_SIZE
	);
	this->returned_buffer[returned_packet_size] = '\0';
	if (sscanf(this->returned_buffer, "[%f,%f]", speed1, speed2) != 2)
	{
		return false;
	}
	return true;
}

const char *Teensy41::get_device_name()
{
	return this->device_name;
}

File: catkin_ws/src/asclinic_pkg/src/drivers/src/teensy41/teensy41.py
import serial


class Teensy41:
    """
    This class is an abstraction to interact with the Teensy to retrieve
    information from the differential encoder.
    """

    def __init__(self, device_name="/dev/ttyACM0", timeout=0):
        self._serial_port = serial.Serial(device_name, timeout=timeout)

    def close_device(self):
        """
        Closes the device.
        """
        if self._serial_port.is_open:
            self._serial_port.close()

    def get_motors_displacement(self):
        """
        Gets motors' displacement without timestamp.
        """
        self._serial_port.write(b"*")
        while len(string_from_teensy) == 0:
            # 50 characters should be enough, but if not, just increase this
            string_from_teensy = self._serial_port.read(50)
        strings_with_braces = string_from_teensy.decode().split(",")
        if strings_with_braces[0][0] != "<" or strings_with_braces[1][-1] != ">":
            error_message = "The data from teensy is corrupted."
            raise Exception(error_message)
        disp1 = int(strings_with_braces[0][1:])
        disp2 = int(strings_with_braces[1][:-1])
        return (disp1, disp2)

    def get_motors_displacement_with_timestamp(self):
        """
        Gets motors' displacemment with timestamp.
        """
        self._serial_port.write(b"t")
        string_from_teensy = ""
        while len(string_from_teensy) == 0:
            # 50 characters should be enough, but if not, just increase this
            string_from_teensy = self._serial_port.read(50)
        strings_with_braces = string_from_teensy.decode().split(",")
        if strings_with_braces[0][0] != "{" or strings_with_braces[-1][-1] != "}":
            error_message = "The data from teensy is corrupted."
            raise Exception(error_message)
        disp1 = int(strings_with_braces[0][1:])
        disp2 = int(strings_with_braces[1])
        timestamp = int(strings_with_braces[-1][:-1])
        return (disp1, disp2, timestamp)

    def get_motors_speed_rpm(self):
        """
        Gets motors' speed in RPM.
        """
        self._serial_port.write(b"s")
        while len(string_from_teensy) == 0:
            # 50 characters should be enough, but if not, just increase this
            string_from_teensy = self._serial_port.read(50)
        strings_with_braces = string_from_teensy.decode().split(",")
        if strings_with_braces[0][0] != "[" or strings_with_braces[1][-1] != "]":
            error_message = "The data from teensy is corrupted."
            raise Exception(error_message)
        speed1 = int(strings_with_braces[0][1:])
        speed2 = int(strings_with_braces[1][:-1])
        return (speed1, speed2)

    @property
    def device_name(self):
        """
        Returns the device's name.
        """
        return self._serial_port.name
